### PR TITLE
Add thread-safe to the library

### PR DESCRIPTION
**Pull request by jchen293**
# Description

This PR adds thread-safe to the EasyPost Java client library, as well as some other minor improvements

- A new EasyPostClient object that has apiKey, connectionTimeout, and readTimeout. The API key parameter is always required when initializing the object, and timeout parameters are optional.
- Split the current class (e.g Address) into model and service folders, where the model folder has all the getters/setters and properties of the Address class (e.g street1, street2, etc), the service folder will have all the functions belonging to that class. This will prevent the users to use any functions from the response object, and the EasyPostClient object will always be used as the root object for any HTTP request call.
- Update the request functions in the Requestor class, the apiKey parameter will be updated to the EasyPostClient object which already has the apiKey and timeouts configuration. Remove the timeout variables and getters/setters in the Requestor class.
- Update all unit tests and configure the TestUtils file.
- Remove the `merge` function and associate helper functions because it's no longer needed as we decided to return the response value directly instead of updating the object.
- AppEngine timeout is now set to the connection timeout.
- Remove an invalid function `User.create()` since creating a child-user always requires a name.
- Move helper functions (instanceUrl, classUrl) to the `Utilities` file from `EasyPostResource` because model classes no longer use it.

# Testing

Re-recorded current unit tests
Add a unit test for EasyPostClient

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

**Review by nwithan8 — CHANGES_REQUESTED** [review]
Please update the examples in the README

**Comment by nwithan8** [comment]
Have we accounted for beta and API versioning in this? We can kick that down the road if needed.

**Comment by jchen293** [comment]
> Have we accounted for beta and API versioning in this? We can kick that down the road if needed.

It's not part of thread-safe PR because I would like to keep the PR focusing on thread-safe only. Once we have another Beta endpoint available for us, we will make a separate PR

**Review by nwithan8 — APPROVED** [review]

File: CHANGELOG.md
Comment (jchen293) [reply CHANGELOG.md]: resolved, good catch!

**Review by nwithan8 — APPROVED** [review]

File: src/main/java-templates/com/easypost/EasyPost.java
Comment (Justintime50) [inline src/main/java-templates/com/easypost/EasyPost.java:35]: Question: Do we want to move these API base strings elsewhere? The client object or the Constant file instead?

File: src/main/java/com/easypost/http/Requestor.java
Comment (Justintime50) [inline src/main/java/com/easypost/http/Requestor.java:469]: This check should happen higher up the stack - you can't create a client object without an API key, so this check shouldn't be necessary and can be removed (it's a duplicate).

File: src/main/java/com/easypost/model/CreditCard.java
Comment (Justintime50) [inline src/main/java/com/easypost/model/CreditCard.java:1]: Let's completely remove the `CreditCard` class and functionality. I know it will be a "version early", but it was added by accident and should be super low impact since it was a one-off. Let's just call it out in the CHANGELOG please.

File: src/main/java/com/easypost/model/PaymentMethodObject.java
Comment (Justintime50) [inline src/main/java/com/easypost/model/PaymentMethodObject.java:175]: I hate that Java doesn't have a scope where you can use a function in your own package but block it from outsiders just because it's in a different folder...

File: src/main/java/com/easypost/service/CreditCardService.java
Comment (Justintime50) [inline src/main/java/com/easypost/service/CreditCardService.java:1]: As mentioned on the model, let's remove the `CreditCard` class and logic completely and call it out in the CHANGELOG.

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (Justintime50) [inline src/main/java/com/easypost/service/EasyPostClient.java:8]: Sort this list please 😉

File: src/test/java/com/easypost/BatchTest.java
Comment (Justintime50) [inline src/test/java/com/easypost/BatchTest.java:199]: Hmm, the ID of an object should always be the first parameter. Param ordering should be tied to the flow. We'll always need to know the object in question (Batch) and then we do stuff to it (add/remove shipments). Other params may come and go but if we are going this route, the ID will always be required.

This is true of all the previous instance based functions. You'll want to check them all.

File: src/test/java/com/easypost/ThreadTest.java
Comment (Justintime50) [inline src/test/java/com/easypost/ThreadTest.java:5]: I'm not seeing a thread test that spins up multiple threads with various API keys and gets the API keys out of those client objects to ensure they didn't get mixed up, can we please add that here?

**Review by Justintime50 — COMMENTED** [review]
I didn't check the rest of the tests (checked a couple), will do that once the rest of this feedback is addressed.

File: src/main/java-templates/com/easypost/EasyPost.java
Comment (nwithan8) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: Jack and I discussed this yesterday. I know in C# we wanted to provide a way for users to override the base URL. Do we still want that here as well?

Comment (jchen293) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: This is currently being used in the Utilities file `singleClassURL` function. Should we allow the user to override the base URL?

File: src/main/java/com/easypost/http/Requestor.java
Comment (jchen293) [reply src/main/java/com/easypost/http/Requestor.java:469]: Users could initialize a EasyPostClient object with a `null` or `empty` string, this check statement is needed in case a user pass in an invalid `API Key` when creating EasyPostClient

Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: The null check should happen (and throw an error) during Client construction.

Comment (jchen293) [reply src/main/java/com/easypost/http/Requestor.java:469]: Yea, we could. But the user will need to add an exception handler in the EasyPostClient initialization method. Any thought?

File: src/main/java-templates/com/easypost/EasyPost.java
Comment (Justintime50) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: Oh yeah, yes please. It should be an optional param to the Client object. Remove the `/v2` from the base URL since that's the version and make sure the version is a private prop of the Client so only we can set it

File: src/main/java/com/easypost/http/Requestor.java
Comment (Justintime50) [reply src/main/java/com/easypost/http/Requestor.java:469]: API key is required, so don't let users create a Client without one (don't provide that overload)

Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: If it throws an error because the user did something incorrectly, it's on the user to account for doing things incorrectly, you know?

File: src/test/java/com/easypost/BatchTest.java
Comment (jchen293) [reply src/test/java/com/easypost/BatchTest.java:199]: Not sure if I follow, but what is the benefit of swapping `param` and `id` as both parameters are required?

File: src/test/java/com/easypost/ThreadTest.java
Comment (jchen293) [reply src/test/java/com/easypost/ThreadTest.java:5]: This is the old unit test, there is a test to ensure API keys do not get mixed up between multiple client objects in `EasyPostTest.java`.

File: src/test/java/com/easypost/BatchTest.java
Comment (Justintime50) [reply src/test/java/com/easypost/BatchTest.java:199]: Consistency across all functions like this. ID will always be required for these previous instance methods whereas some of these will have optional params. If ID comes first, then it's always the same order for all functions.

File: src/test/java/com/easypost/ThreadTest.java
Comment (Justintime50) [reply src/test/java/com/easypost/ThreadTest.java:5]: If so, the thread test should be placed in the thread test test file, don't you agree?

File: src/main/java-templates/com/easypost/EasyPost.java
Comment (jchen293) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: This will be the fourth parameter in the EasyPostClient constructor, are we allowing users to override the default `/v2` version?

Comment (Justintime50) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: No, I said:

> make sure the version is a private prop of the Client so only we can set it

Comment (jchen293) [reply src/main/java-templates/com/easypost/EasyPost.java:35]: Ok, no optional param should be given in the constructor.

File: src/test/java/com/easypost/ThreadTest.java
Comment (jchen293) [reply src/test/java/com/easypost/ThreadTest.java:5]: sure

File: src/test/java/com/easypost/BatchTest.java
Comment (jchen293) [reply src/test/java/com/easypost/BatchTest.java:199]: Just looked at Go, indeed we the ID comes first before any other parameters

**Comment by nwithan8** [comment]
`apiBase` should be, e.g., "https://api.easypost.com", and should be customizable in the Client constructor.

`apiVersion` should be, e.g., "v2" and is set internally (end-user doesn't know about it).

**Comment by jchen293** [comment]
> `apiBase` should be, e.g., "https://api.easypost.com", and should be customizable in the Client constructor.
> 
> `apiVersion` should be, e.g., "v2" and is set internally (end-user doesn't know about it).

If `apiBase` is customizable, what if the user overrides with the wrong API base URL?

**Comment by nwithan8** [comment]
> > `apiBase` should be, e.g., "https://api.easypost.com", and should be customizable in the Client constructor.
> > 
> > `apiVersion` should be, e.g., "v2" and is set internally (end-user doesn't know about it).
> 
> If `apiBase` is customizable, what if the user overrides with the wrong API base URL in the Client constructor?
> 

Then their requests will probably fail. Again, it's on the end user to account for that if they are going to utilize "advanced features" like that.

**Comment by jchen293** [comment]
> > > `apiBase` should be, e.g., "https://api.easypost.com", and should be customizable in the Client constructor.
> > > `apiVersion` should be, e.g., "v2" and is set internally (end-user doesn't know about it).
> > 
> > 
> > If `apiBase` is customizable, what if the user overrides with the wrong API base URL in the Client constructor?
> 
> Then their requests will probably fail. Again, it's on the end user to account for that if they are going to utilize "advanced features" like that.

I am not a big fan of this. We probably don't have much `API base`, I think we could use Enum for the `apiBase` optional parameter, this would prevent spelling mistakes like typos from the user input.

**Comment by nwithan8** [comment]
> > > > `apiBase` should be, e.g., "https://api.easypost.com", and should be customizable in the Client constructor.
> > > > `apiVersion` should be, e.g., "v2" and is set internally (end-user doesn't know about it).
> > > 
> > > 
> > > If `apiBase` is customizable, what if the user overrides with the wrong API base URL in the Client constructor?
> > 
> > Then their requests will probably fail. Again, it's on the end user to account for that if they are going to utilize "advanced features" like that.
> 
> I am not a big fan of this. We probably don't have much `API base`, I think we could use Enum for the `apiBase` optional parameter, this would prevent spelling mistakes like typos from the user input.

See how we do it in C#: https://github.com/EasyPost/easypost-csharp/blob/master/EasyPost/Client.cs#L63

**Comment by Justintime50** [comment]
@jchen293 we want to allow users to override the base API for two reasons:

1. We could point the client library at our internal test servers if needed for end to end test something
2. If we ever released a mock API product in the future, users could override the base API to hit that instead (for testing or setup purposes)

It should be a free-hand string users can specify. It will not be documented as the use cases will be slim; however, some of our other libs like Python allow you to override this already and we'd like them all to allow it.

As for the versioning, client libs will always be locked to the most recent version available and this cannot be customized. In theory, whatever base APIs are available will have the same set of API versions available to them.

**Comment by jchen293** [comment]
> @jchen293 we want to allow users to override the base API for two reasons:
> 
> 1. We could point the client library at our internal test servers if needed for end to end test something
> 2. If we ever released a mock API product in the future, users could override the base API to hit that instead (for testing or setup purposes)
> 
> It should be a free-hand string users can specify. It will not be documented as the use cases will be slim; however, some of our other libs like Python allow you to override this already and we'd like them all to allow it.
> 
> As for the versioning, client libs will always be locked to the most recent version available and this cannot be customized. In theory, whatever base APIs are available will have the same set of API versions available to them.

Got it, we probably don't want to expose the `API Base` in the source code by using Enum, I will add another optional parameter for the API base next week.

File: src/main/java/com/easypost/service/ReferralCustomerService.java
Comment (Justintime50) [inline src/main/java/com/easypost/service/ReferralCustomerService.java:226]: I'd prefer we set all these `String.format` items to an intermediary variable so we know what it is and these lines can be much shorter. Thoughts?

If you agree, we should do that throughout the whole project. It just seems crammed with so much text on one line which then wraps to two or three..

Comment (jchen293) [reply src/main/java/com/easypost/service/ReferralCustomerService.java]: I agree, this is resolved.

**Review by Justintime50 — APPROVED** [review]
I thiiink all of my feedback was addressed. This is good work!

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/EasyPostClient.java:173]: We don't need getters for the API version. Make the `apiVersion` property public and final, so it can't be overridden by end-users (only set via the constructor)

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/EasyPostClient.java:11]: Make `apiVersion` public (see below)

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/EasyPostClient.java:12]: Make this final, since it won't change after being set by the constructor.

File: src/main/java/com/easypost/service/BillingService.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/BillingService.java]: Since the API base won't change for different API calls (it's not like one API call is going to go to a different base url than every other API call + the fact that API base can't be changed after construction anyway), we can move API base formatting downstream into the `request` function.

API version can change on a per-call basis (some go to Beta, some go to v2, etc), that should remain here as a parameter.

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/EasyPostClient.java:182]: We don't need getters for the API base. Make the `apiBase` property public and final, so it can't be overridden by end-users (only set via the constructor)

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/EasyPostClient.java:173]: Alternatively, if we are going to keep this getter, let's have it accept a parameter so it can return `v2` by default, but other possible versions (i.e. `beta`) on request.

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [inline src/main/java/com/easypost/service/EasyPostClient.java:105]: Nit: I hate all these overloads. Have an idea for a better way, but won't block on it.

**Review by nwithan8 — CHANGES_REQUESTED** [review]

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:12]: I think this is just our personal preference, I like to use getter/setter for immutable variables in Java to follow the `encapsulation` practice, unless there is a significant advantage by making these two fields `public final` then I am down to make that change.

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:11]: see below answer

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:173]: We should not have an additional parameter in the getter method, this will allow end users to change the apiVersion variable. The reason why the `apiVersion` is not final so we could change it internally in the `EasyPostClient` if the namespace has `beta`, e.g: `client.beta.endshipper.create()`

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/EasyPostClient.java:12]: Fair enough, won't block.

File: src/main/java/com/easypost/service/BillingService.java
Comment (jchen293) [reply src/main/java/com/easypost/service/BillingService.java]: Requestor class has it, in line 429. This function is not using the `classURL()` from the `Utilities` file because the endpoint has a different name from the model class name (Billing), that's why we are passing in the `API base` here in the `String.format`

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:182]: see above answer

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:105]: So am I? It's Java things

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/EasyPostClient.java:182]: At least make the apiBase final

File: src/main/java/com/easypost/service/BillingService.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/BillingService.java]: Not a fan of the classURL() utility if it's going to cause confusing edge cases like this when we DON'T use it. Won't block, but could do with a re-imagining down the line.

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/EasyPostClient.java:173]: I foresee some issues arising when we do need to implement a new beta service down the road with this current design, but we'll kick the can on that for now.

**Review by nwithan8 — APPROVED** [review]

File: src/main/java/com/easypost/service/EasyPostClient.java
Comment (jchen293) [reply src/main/java/com/easypost/service/EasyPostClient.java:182]: I will make this tiny change in next PR so I don't dismiss the approvals

**Pull request merged**
